ReadableStream cancel method not reliably called in Service Worker

youennf opened the pull request:
#### 07ca47f11ff737fbf0d24b30529de01cff412a52
<pre>
ReadableStream cancel method not reliably called in Service Worker
<a href="https://rdar.apple.com/144297119">rdar://144297119</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=286738">https://bugs.webkit.org/show_bug.cgi?id=286738</a>

Reviewed by Per Arne Vollan.

We prevent the fetch response used for the download to be gced when its readable stream body is used to consume data.
This allows the path to cancel the fetch response to reliably work and execute the readable stream cancel source callback.
We cannot take a ref to the fetch response in the callback as the callback may be destroyed in the main thread and this may create a reference cycle.

* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::hasPendingActivity const):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
* Source/WebCore/Modules/streams/ReadableStreamSink.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
(-[ServiceWorkerDownloadDelegate _download:decideDestinationWithSuggestedFilename:completionHandler:]):
((ServiceWorker, ServiceWorkerReadableStreamDownloadCancel)):

Canonical link: <a href="https://commits.webkit.org/291551@main">https://commits.webkit.org/291551@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a9507558f851304860cceb02f28b427e5eb008d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93092 "21 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12644 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98091 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43618 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95142 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21100 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71182 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28599 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96094 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9744 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84245 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51510 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9437 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1878 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42931 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79728 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1860 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100118 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14778 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80205 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20396 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79507 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19781 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24051 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1375 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13239 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20128 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25304 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19815 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23275 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21556 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -182,6 +184,20 @@ - (void)webView:(WKWebView *)webView decidePolicyForNavigationResponse:(WKNaviga
 }
 @end
 
+@interface ServiceWorkerDownloadDelegate : NSObject <_WKDownloadDelegate>
```
Can `WKDownloadDelegate` be used here?

youennf replied:
Good point, I removed SPI use by using another navigation delegate and calling WKDownload cancel.

pvollan reviewed (approved):
R=me.

youennf committed:
ReadableStream cancel method not reliably called in Service Worker
rdar://144297119
https://bugs.webkit.org/show_bug.cgi?id=286738

Reviewed by Per Arne Vollan.

We prevent the fetch response used for the download to be gced when its readable stream body is used to consume data.
This allows the path to cancel the fetch response to reliably work and execute the readable stream cancel source callback.
We cannot take a ref to the fetch response in the callback as the callback may be destroyed in the main thread and this may create a reference cycle.

* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::hasPendingActivity const):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
* Source/WebCore/Modules/streams/ReadableStreamSink.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
(-[ServiceWorkerDownloadDelegate _download:decideDestinationWithSuggestedFilename:completionHandler:]):
((ServiceWorker, ServiceWorkerReadableStreamDownloadCancel)):

Canonical link: https://commits.webkit.org/291551@main
